### PR TITLE
[RFC] options: make 'highlight' read-only

### PR DESCRIPTION
**Pull request by bfredl**
broken out of #6700

### CODE REVIEW

File: runtime/doc/vim_diff.txt [hunk 1]
```diff
@@ -310,6 +310,7 @@ Other options:
   'esckeys'
   'guioptions' "t" flag was removed
   *'guipty'* (Nvim uses pipes and PTYs consistently on all platforms.)
+  'highlighht' (the builtin default group names are allways used)
```
Comment (justinmk) [inline runtime/doc/vim_diff.txt]: typos

Comment (bfredl) [inline runtime/doc/vim_diff.txt]: fixed!

File: src/nvim/option.c [hunk 2]
```diff
@@ -2538,11 +2538,11 @@ did_set_string_option (
...
     }
-  }
-  /* 'highlight' */
-  else if (varp == &p_hl) {
-    if (highlight_changed() == FAIL)
-      errmsg = e_invarg;        /* invalid flags */
+  } else if (varp == &p_hl) {
+    // 'highlight'
+    if (strcmp((char *)(*varp), HIGHLIGHT_INIT) != 0) {
+      errmsg = e_unsupportedoption;
```
Comment (justinmk) [inline src/nvim/option.c]: Could we give an error message like `E519: Use the :highlight command instead.`

File: test/functional/ex_cmds/encoding_spec.lua [hunk 3]
```diff
@@ -25,7 +25,7 @@ describe('&encoding', function()
     clear('--cmd', 'set enc=latin1')
```
Comment (justinmk) [inline test/functional/ex_cmds/encoding_spec.lua]: The "after setup" and "before startup" test descriptions need to be updated. 

Also noticed that `e_afterinit` is not used, though it should be for 'termguicolors'.

File: src/nvim/option.c (hunk 2 above)
Comment (bfredl) [inline src/nvim/option.c]: Is this needed here? Only in very special cases would someone needed to change this option, like someone forced to use a vim build without +syntax. Compare with `&encoding`, we do not say "change fileencoding to edit files in different encodings" in the error message.

Comment (justinmk) [inline src/nvim/option.c]: Good point. I was imagining someone with a config they copied or an old plugin.

File: test/functional/ex_cmds/encoding_spec.lua (hunk 3 above)
Comment (bfredl) [inline test/functional/ex_cmds/encoding_spec.lua]: e_afterinit should not be used, as there is no difference in behavior for and after vimenter.

and 'termguicolor' should be fixed instead. Add `ui_call_gui_option_changed` and include 'termguicolor' amoung the gui options.

Comment (justinmk) [inline test/functional/ex_cmds/encoding_spec.lua]: I mean e_afterinit is not used anywhere.

Comment (bfredl) [inline test/functional/ex_cmds/encoding_spec.lua]: Yes, but that is not a problem if we currently don't have any option of that kind. :)

**Comment by teto** [comment]
Seems to make sense to enforce defaults so that plugins can predict the hl names to override.
:ownsyntax seems to cover the usecase. Is there any functionality lost by making 'highlight' RO ?
I tried a `"set highlight=" site::https://github.com` on google to see if it was used but couldn't make anything of the results.

Fun fact, while looking at creating the "Normal" hl with id 0 in https://github.com/neovim/neovim/issues/6738 I discovered that 'highlight' was responsible for the order of id->hl group.

**Comment by bfredl** [comment]
:ownsyntax is othogonal; this cover ui highlighting, not syntax highlighting.

Isn't ordering determined by the HLF_ enum? Creation of the groups should run before any config or syntax definitions create other groups, so the id of these groups should be stable (still, it is not a good idea to hardcode them outside nvim core, better use the names).

**Comment by teto** [comment]
Nope it's the parsing of 'highlight' that creates the group first
```
(gdb) bt
Reading in symbols for ../src/nvim/option.c...done.
#0  syn_add_group (name=0x7ffff682f180 "SpecialKey") at ../src/nvim/syntax.c:7261
#1  0x00005555557e64e7 in syn_check_group (
    pp=0x5555558e773a "SpecialKey,~:EndOfBuffer,z:TermCursor,Z:TermCursorNC,@:NonText,d:Directory,e:ErrorMsg,i:IncSearch,l:Search,m:MoreMsg,M:ModeMsg,n:LineNr,N:CursorLineNr,r:Question,s:StatusLine,S:StatusLineNC,c:VertSpli"..., len=10) at ../src/nvim/syntax.c:7244
#2  0x00005555557e6b63 in highlight_changed () at ../src/nvim/syntax.c:7457
#3  0x0000555555731200 in didset_options2 () at ../src/nvim/option.c:2127
#4  0x000055555572d8bb in set_init_1 () at ../src/nvim/option.c:780
#5  0x00005555556d3242 in early_init () at ../src/nvim/main.c:212
#6  0x00005555556d32f4 in main (Reading in symbols for ../sysdeps/x86/libc-start.c...done.
argc=4, argv=0x7fffffffd8d8) at ../src/nvim/main.c:239
(gdb) c
```

**Comment by bfredl** [comment]
Seems you're right, but more precisely the default for `'highlight'` (which almost identical to HLF_ order), when vimrc or otherwise can change the value the default has already been parsed.

**Comment by justinmk** [comment]
Should the order of HIGHLIGHT_INIT be changed to match `hlf_names`? Just need to swap `SignColumn` and `Conceal`. Also, `VisualNOS` should be removed from `hlf_names`, we removed it long ago.

File: src/nvim/syntax.c [hunk 4]
```diff
@@ -7314,100 +7314,27 @@ static void highlight_attr_set_all(void)
...
-          else if (hlf == (int)HLF_S)
-            id_S = syn_get_final_id(id);
-          break;
-        default:    return FAIL;
-        }
-      }
-      highlight_attr[hlf] = attr;
-
-      p = skip_to_option_part(p);           /* skip comma and spaces */
+  for (int hlf = 0; hlf < (int)HLF_COUNT; hlf++) {
```
Comment (justinmk) [inline src/nvim/syntax.c]: Could use a brief comment here:

    // Reset to the builtin highlights.

File: src/nvim/syntax.c [hunk 5]
```diff
@@ -7314,100 +7314,27 @@ static void highlight_attr_set_all(void)
  */
 int highlight_changed(void)
```
Comment (justinmk) [inline src/nvim/syntax.c]: The doc has some outdated parts now. 

Also `get_highlight_default()` is not used anywhere now.

**Comment by bfredl** [comment]
We do plan on reimplement autoselect, so visualNOS should be kept.

**Comment by justinmk** [comment]
In the merge I adjusted the docs. Also added 'winhighlight' to vim_diff, and removed "VisualNOS" from deprecated.txt.

**Pull request merged**